ovs/extensions/storageserver/storagedriverconfig/base.py

```diff
@@ -0,0 +1,203 @@
+# Copyright (C) 2016 iNuron NV
+#
+# This file is part of Open vStorage Open Source Edition (OSE),
+# as available from
+#
+#      http://www.openvstorage.org and
+#      http://www.openvstorage.com.
+#
+# This file is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
+# as published by the Free Software Foundation, in version 3 as it comes
+# in the LICENSE.txt file of the Open vStorage OSE distribution.
+#
+# Open vStorage is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY of any kind.
+"""
+Base storagedriver config
+Exposes methods for all sections of a complete storagedriver config
+"""
+from subprocess import check_output
+
+
+class VolumeDriverConfigOption(object):
+    """
+    Config option read out by 'volumedriver_fs', '--config-help-markdown'
+    """
+    _options_by_component_and_key = {}
+
+    def __init__(self, component, key, default_value, dynamically_reconfigurable, remarks):
+        #  type: (str, str, any, bool, str) -> None
+        """
+        Initialize a new option
+        Format of the command is '| component | key | default value | dynamically reconfigurable | remarks |'
+        :param component: Volumedriver component
+        :type component: str
+        :param key: Option key
+        :type key: str
+        :param default_value: Default value
+        :param dynamically_reconfigurable:
+        :param remarks:
+        """
+        self.component = component
+        self.key = key
+        self.default_value = default_value
+        self.dynamically_reconfigurable = dynamically_reconfigurable
+        self.remarks = remarks
+
+    @classmethod
+    def get_all_options(cls):
+        # type: () -> List[VolumeDriverConfigOption]
+        """
+        Get all possible options
+        :return: A list with all options
+        :rtype: List[VolumeDriverConfigOption]
+        """
+        if not cls._options_by_component_and_key:
+            cls._cache_options()
+        options = []
+        for component, component_keys_options in cls._options_by_component_and_key.iteritems():
+            for component_key, option in component_keys_options.iteritems():
+                options.append(option)
+        return options
+
+    @classmethod
+    def get_option_by_component_and_key(cls, component, key):
+        # type: (str, str) -> VolumeDriverConfigOption
+        """
+        Retrieve an option by component and key
+        Returns None when no option was found
+        :param component: Component to retrieve
+        :type component: str
+        :param key: Key to retrieve from component
+        :type key: str
+        :return: Option if found else none
+        :rtype: VolumeDriverConfigOption
+        """
+        if not cls._options_by_component_and_key:
```

Comment:
Instead of checking every time, wouldn't always calling a function `_validate_cache` or sth and differentiating in behaviour there, be more readable?